namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala

```diff
@@ -32,80 +30,84 @@ class DnsSrvNamer(
+    val state = Var.async[Activity.State[List[Address]]](Activity.Pending) { update =>
+
+      def doUpdate() = {
+        val lookup = new DNS.Lookup(dsnsrvRecord, DNS.Type.SRV, DNS.DClass.IN)
+        lookup.setResolver(resolver)
+        Stat.time(latency)(lookup.run())
+        lookup.getResult match {
+          case DNS.Lookup.HOST_NOT_FOUND | DNS.Lookup.TYPE_NOT_FOUND =>
+            log.trace("no results for %s", dsnsrvRecord)
+            failure.incr()
+            update.update(Activity.Failed(new Exception("Host not found")))
+          case DNS.Lookup.SUCCESSFUL =>
+            val answers = Option(lookup.getAnswers).getOrElse(Array.empty)
+            val srvRecords = answers.flatMap {
+              case srv: DNS.SRVRecord => try {
+                val inetAddress = InetAddress.getByName(srv.getTarget.toString())
+                Some(Address(new InetSocketAddress(inetAddress, srv.getPort)))
+              } catch {
+                case _: UnknownHostException =>
+                  log.warning(s"srv lookup of $dsnsrvRecord returned unknown host ${srv.getTarget}")
+                  badHosts.incr()
+                  None
+              }
+              case _ => None
+            }
+            if (srvRecords.isEmpty) {
+              // valid DNS entry, but no instances.
+              // return NameTree.Neg because NameTree.Empty causes requests to fail,
+              // even in the presence of load-balancing (NameTree.Union) and fail-over (NameTree.Alt)
+              log.trace("empty response for %s", dsnsrvRecord)
+              zeroResults.incr()
+              update.update(Activity.Failed(new Exception(s"empty response for $dsnsrvRecord")))
+            } else {
+              log.trace("got %d results for %s", srvRecords.length, dsnsrvRecord)
+              success.incr()
+              update.update(Activity.Ok(srvRecords.toList))
             }
-            case _ => None
-          }
-          if (srvRecords.isEmpty) {
-            // valid DNS entry, but no instances.
-            // return NameTree.Neg because NameTree.Empty causes requests to fail,
-            // even in the presence of load-balancing (NameTree.Union) and fail-over (NameTree.Alt)
-            log.trace("empty response for %s", address)
-            zeroResults.incr()
-            NameTree.Neg
-          } else {
-            log.trace("got %d results for %s", srvRecords.length, address)
-            success.incr()
-            NameTree.Leaf(Name.Bound(Var.value(Addr.Bound(srvRecords: _*)), id, residual))
-          }
-        case code =>
-          val msg = s"unexpected result: $code for $address: ${lookup.getErrorString}"
-          log.error(msg)
-          throw new IOException(msg)
+          case code =>
+            val msg = s"unexpected result: $code for $dsnsrvRecord: ${lookup.getErrorString}"
+            log.error(msg)
+            update.update(Activity.Failed(new Exception(msg)))
+        }
+      }
+
+      val _ = doUpdate()
```

Comment:
if you don't use the return value, maybe give the method a type of `doUnit(): Unit`?